bitcoin-str/Cargo.toml

```diff
@@ -0,0 +1,41 @@
+[package]
+name = "bitcoin-str"
+version = "0.27.0"
+authors = ["Andrew Poelstra <apoelstra@wpsoftware.net>"]
+license = "CC0-1.0"
+homepage = "https://github.com/rust-bitcoin/rust-bitcoin/"
+repository = "https://github.com/rust-bitcoin/rust-bitcoin/"
+documentation = "https://docs.rs/bitcoin/"
+description = "The string encodings used by rust-bitcoin (base58, base58check, bech32, bech32m, base64, hex)."
+keywords = [ "crypto", "bitcoin" ]
+readme = "https://github.com/rust-bitcoin/rust-bitcoin/README.md"
+
+# Please don't forget to add relevant features to docs.rs below
+[features]
+default = ["std"]
+unstable = []
+base64 = ["base64-compat"]
+use-serde = ["serde", "bitcoin_hashes/serde"]
+
+# At least one of std, no-std must be enabled.
+#
+# The no-std feature doesn't disable std - you need to turn off the std feature for that by disabling default.
+# Instead no-std enables additional features required for this crate to be usable without std.
+# As a result, both can be enabled without conflict.
+std = ["bitcoin_hashes/std", "bech32/std"]
+no-std = ["core2/alloc", "bitcoin_hashes/alloc"]
+
+[package.metadata.docs.rs]
+features = [ "std", "secp-recovery", "base64", "rand", "use-serde", "bitcoinconsensus" ]
+rustdoc-args = ["--cfg", "docsrs"]
+
+[dependencies]
+bech32 = { version = "0.8.1", default-features = false }
+bitcoin_hashes = { version = "0.10.0", default-features = false }
```

Comment:
Do we need the hashes as well or are we just taking `hex` from this? Would be nice to move `hex` into `bitcoin-str` and have `bitcoin-hashes` depend on `bitcoin-str`. AFAIK, `hex` only lives in `bitcoin-hashes` because `bitcoin-hashes` was already a leaf dependency.

Reply:
concept ACK that dependency inversion.

Yes, the hex stuff was put into `bitcoin-hashes` purely because it was the smallest crate in the ecosystem. (I might've put hex into `bech32` if it'd been around at the time :P.)

Reply:
There are three functions in `base58` module that depend on hashing, e.g.
```
/// Obtain a string with the base58check encoding of a slice
/// (Tack the first 4 256-digits of the object's Bitcoin hash onto the end.)
pub fn check_encode_slice(data: &[u8]) -> String {
    let checksum = sha256d::Hash::hash(data);
    encode_iter(
        data.iter()
            .cloned()
            .chain(checksum[0..4].iter().cloned())
    )
}
```

Reply:
Why we don't use the `hex` crate?

Reply:
Because it's a 3rd-party dep and its functionality can be re-implemented in ten lines of code.

We have had problems in the past with these things (e.g. `base64` has caused us more trouble than it would've been to NIH hex *and* base64).

Reply:
I [once suggested](https://github.com/Kixunil/loptos/issues/1#issuecomment-888896092) to use base64 feature from bitcoin crate and this random guy :) said `base64` was already in his crate as a transitive dep so he didn't want to add another base64 impl which is a reasonable point I think. 

Was the problem you mention about `base64` related to MSRV or something else?

Reply:
Yes, related to MSRV, and a quite hostile response from the maintainer when we tried to fix it.

Reply:
> I was hoping to convey that with the comment in `bitcoin-str/src/lib.rs`
> ```
> pub mod hex {
>     //! Re-exports hex de/serialization code from `bitcoin_hashes`.
>     //!
>     // There is a chicken and the egg problem with putting _all_ the en/decoding code in a single crate.
>     // Hashing cannot be separated totally from en/decoding because base58 -> hashing -> hex.
> ```
> 
> Or is there a way around this?

I can see the following options:

1. Make `bitcoin_hashes` _only_ do hashing with APIs that go from byte -> byte, removing the hex code from it
2. Move `bitcoin_hashes` into `bitcoin-str`. Not entirely correct but if we want to tightly couple them with hex then this issue would be resolved
3. Duplicate the hex-code across `bitcoin_hashes` and `bitcoin-str` and make it an implementation detail in `bitcoin_hashes`

(3) might overall be the cleanest approach regarding exposed APIs and crate structure. The hex code needed to parse and represent hashes is quite specific whereas if we want to expose hex encoding in `bitcoin-str` for general use then there are a lot more use-cases to support. I am thinking lazy iterators, stack-allocated encoding, etc.

Reply:
I am leaning toward option (3), simply because hex is such a simple encoding that I'm not too bothered by its duplication.

Reply:
4. Move the code to a separate crate.
5. Discuss with the maintainer about MSRV and API stabilization - not everyone is hostile

I think 4 short term and 5 long-term would be nice but can also accept 3 since it's simple. (TBH, I also reimplemented it in one of my crates.)

Reply:
I already made a separate bitcoin_hex crate more than a year ago after we had a similar discussion, but its adoption here ended up nowhere and I archived it.

Reply:
Here the PR, closed it just three days ago: https://github.com/rust-bitcoin/bitcoin_hashes/pull/91

Reply:
After reading [your argument](https://github.com/rust-bitcoin/rust-bitcoin/issues/467#issuecomment-687018559) I no longer feel comfortable with accepting option 3.

I believe we should really revive it. But could the crate be without "bitcoin" in its name or docs? I'd like it being neutral general purpose 1.36-compatible crate. `hex-compat` similarly to `base64-compat` sounds sensible.

Reply:
There is already `amplify_num` crate I made for this purpose,  when work on `bitcoin_hex` has stalled. I will be ok with moving it to this org and even renaming it.

Reply:
I'm unconvinced that "changes will happen" to hex encoding code which require synchronized updates, but concept ACK having a separate hex-compat crate.

Agree that it shouldn't have `bitcoin` anywhere in the name.

I'm also OK in concept with talking to the `hex` maintainer but I have no interest in doing this myself.

Reply:
> I'm unconvinced that "changes will happen" to hex encoding code which require synchronized updates, but concept ACK having a separate hex-compat crate.

That was my thinking as well. Very likely, the hex-encoding needs within `bitcoin-hashes` are not going to change any time soon and hence we might be able to reduce it down to 2-3 private functions that have exactly the signatures we need because the don't need to account for any other usage.

Contrary to that, a public `hex` module within `bitcoin-str` will likely need to offer much more flexible APIs, see more usage and thus require more maintenance.

That was the thinking when I proposed option 3 above. I am not fussed about extracting a dedicated crate though.

Reply:
OK, I've mentioned there that I wish for MSRV to be no newer than 1.41.1 and noticed they also have 1.0 milestone which is very nice. I can see that the code being short and unlikely to change can be a good argument for duplication being fine but still prefer a separate crate. It could be a fork of `hex` with changes to make it compilable on 1.41.1.